apiserver/pkg/server/ray_job_submission_service_server_test.go

```diff
@@ -0,0 +1,221 @@
+package server
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubernetesfake "k8s.io/client-go/kubernetes/fake"
+
+	"github.com/ray-project/kuberay/apiserver/pkg/client"
+	"github.com/ray-project/kuberay/apiserver/pkg/manager"
+	"github.com/ray-project/kuberay/apiserver/pkg/util"
+	api "github.com/ray-project/kuberay/proto/go_client"
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	"github.com/ray-project/kuberay/ray-operator/controllers/ray/utils"
+	fakeclientset "github.com/ray-project/kuberay/ray-operator/pkg/client/clientset/versioned/fake"
+)
+
+func TestGetRayClusterURL(t *testing.T) {
+	ctx := context.Background()
+
+	namespace := "test-namespace"
+	clusterName := "test-raycluster"
+
+	validCluster := rayv1.RayCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterName,
+			Namespace: namespace,
+			Labels: map[string]string{
+				util.KubernetesManagedByLabelKey: util.ComponentName,
+			},
+		},
+		Status: rayv1.RayClusterStatus{
+			State: rayv1.Ready,
+		},
+		Spec: rayv1.RayClusterSpec{
+			HeadGroupSpec: rayv1.HeadGroupSpec{
+				Template: corev1.PodTemplateSpec{
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  "test",
+								Image: "test",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	tests := []struct {
+		rayCluster          *rayv1.RayCluster
+		rayEvent            *corev1.Event
+		name                string
+		expectedURL         string
+		expectedErrorString string
+	}{
+		{
+			name:                "Get URL from a valid cluster",
+			rayCluster:          &validCluster,
+			expectedURL:         clusterName + "-head-svc." + namespace + ".svc.cluster.local:8265",
+			expectedErrorString: "",
+		},
+		{
+			name: "Get URL from a cluster with missing name",
+			rayCluster: func() *rayv1.RayCluster {
+				newCluster := validCluster
+				newCluster.ObjectMeta.Name = ""
+				return &newCluster
+			}(),
+			expectedURL:         "",
+			expectedErrorString: "Cluster name is empty. Please specify a valid value.",
+		},
+		{
+			name: "Get URL from a cluster with missing namespace",
+			rayCluster: func() *rayv1.RayCluster {
+				newCluster := validCluster
```

Comment:
is there a way to make deepcopy?
My concern is, it's possible some fields in the protobuf message is pointer, so the original valid pb message is changed during tests.